californium-core/src/test/java/org/eclipse/californium/core/test/lockstep/LockstepEndpoint.java

```diff
@@ -316,6 +384,42 @@ public void check(Message message) {
 			for (Expectation<Message> expectation:expectations)
 				expectation.check(message);
 		}
+		
+		@Override
+		public void go() throws Exception {
+			if (null != multi) {
+				add(multi);
+				return;
+			}
+			
+			RawData raw = incoming.poll(2, TimeUnit.SECONDS); // or take()?
+			Assert.assertNotNull("Did not receive a message (but nothing)", raw);
+
+			Message msg = parser.parseMessage(raw);
+			msg.setSource(raw.getAddress());
+			msg.setSourcePort(raw.getPort());
+			go(msg);
+		}
+		
+		public String toString() {
+			StringBuffer result = new StringBuffer("{");
+			for (Expectation<Message> expectation:expectations) {
+				String info = expectation.toString();
+				if (!info.isEmpty()) {
+					result.append(info).append(",");
+				}
+			}
+			int end = result.length() - 1;
+			if (0 <= end && ',' == result.charAt(end)) {
+				result.setLength(end);
+			}
+			result.append("}");
+			return result.toString();
+		}
+
+		public abstract void go(Message msg) throws Exception;
+
+		public abstract void add(MultiMessageExpectation multi);
```

Comment:
Ok, now I see what you mean. I think you're right, we should keep it like this in order to prevent ugly instanceof checks ...